File: kt-src/check.kt
#include "types.kth"
// check.kt — Static analysis: escape analysis for stack locals

let mut check_locals: [&str; 64]
let mut check_nlocals: i32 = 0

fn collect_locals(block: *AstNode) {
    if block == 0 { return }
    if block.kind != NodeKind::BLOCK { return }
    let mut i: i32 = 0
    while i < block.d1 {
        let stmt: *AstNode = block.d0[[i]]
        if stmt.kind == NodeKind::LET {
            // d2 > 0 means buffer
            if stmt.d2 > 0 {
                if check_nlocals < 64 {
                    check_locals[[check_nlocals]] = stmt.d0
                    check_nlocals = check_nlocals + 1
                }
            }
        }
        i = i + 1
    }
}

fn is_local(name: &str) -> i32 {
    let mut i: i32 = 0
    while i < check_nlocals {
        if streq(check_locals[[i]], name) {
            return 1
        }
        i = i + 1
    }
    return 0
}

fn check_return_expr(expr: *AstNode, ret_loc: SrcLoc) {
    if expr == 0 { return }
    if expr.kind == NodeKind::IDENT && is_local(expr.d0) {
        error_at(ret_loc, "returning pointer to stack-allocated variable")
    }
}

fn check_block(block: *AstNode) {
    if block == 0 { return }
    if block.kind != NodeKind::BLOCK { return }
    let mut i: i32 = 0
    while i < block.d1 {
        let stmt: *AstNode = block.d0[[i]]
        if stmt.kind == NodeKind::RETURN {
            check_return_expr(stmt.d0, stmt.loc)
        }
        i = i + 1
    }
}

fn check_fn(f: *AstNode) {
    let body: *AstNode = f.d1
    if body == 0 { return }
    if body.kind != NodeKind::BLOCK { return }

    check_nlocals = 0
    collect_locals(body)
    if check_nlocals == 0 { return }

    check_block(body)
}

fn check_escape(program: *AstNode) {
    let mut i: i32 = 0
    while i < program.d1 {
        let decl: *AstNode = program.d0[[i]]
        if decl.kind == NodeKind::FN_DEF {
            check_fn(decl)
        } else if decl.kind == NodeKind::ANNOTATION {
            let child: *AstNode = decl.d2
            if child != 0 && child.kind == NodeKind::FN_DEF {
                check_fn(child)
            }
        }
        i = i + 1
    }
}

File: src/check.c
#include "check.h"
#include <string.h>

#define MAX_LOCALS 64

static const char *locals[MAX_LOCALS];
static int nlocals;

static void collect_locals(AstNode *block) {
    if (!block || block->kind != NODE_BLOCK) return;
    for (int i = 0; i < block->block.nstmts; i++) {
        AstNode *stmt = block->block.stmts[i];
        if (stmt->kind == NODE_LET && stmt->let.is_buffer) {
            if (nlocals < MAX_LOCALS)
                locals[nlocals++] = stmt->let.name;
        }
    }
}

static int is_local(const char *name) {
    for (int i = 0; i < nlocals; i++) {
        if (strcmp(locals[i], name) == 0)
            return 1;
    }
    return 0;
}

static void check_return_expr(AstNode *expr, SrcLoc ret_loc) {
    if (!expr) return;
    if (expr->kind == NODE_IDENT && is_local(expr->ident.name)) {
        error_at(ret_loc,
            "returning pointer to stack-allocated '%s' — "
            "memory will be invalid after function returns",
            expr->ident.name);
    }
}

static void check_block(AstNode *block) {
    if (!block || block->kind != NODE_BLOCK) return;
    for (int i = 0; i < block->block.nstmts; i++) {
        AstNode *stmt = block->block.stmts[i];
        if (stmt->kind == NODE_RETURN) {
            check_return_expr(stmt->ret.expr, stmt->loc);
        }
    }
}

static void check_fn(AstNode *fn) {
    AstNode *body = fn->fn_def.body;
    if (!body || body->kind != NODE_BLOCK) return;

    nlocals = 0;
    collect_locals(body);
    if (nlocals == 0) return;

    check_block(body);
}

void check_escape(AstNode *program) {
    for (int i = 0; i < program->program.ndecls; i++) {
        AstNode *decl = program->program.decls[i];
        if (decl->kind == NODE_FN_DEF) {
            check_fn(decl);
        } else if (decl->kind == NODE_ANNOTATION &&
                   decl->annotation.child &&
                   decl->annotation.child->kind == NODE_FN_DEF) {
            check_fn(decl->annotation.child);
        }
    }
}
